### PR TITLE
(BOLT-956) Add wait_until_available function

### PR DESCRIPTION
**Pull request by MikaelSmith**
Add the `wait_until_available` function to wait for all targets to be available for connections. Accepts a description, wait_time in seconds (default 120), and retry_interval in seconds (default 1).

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Needs tests.

File: bolt-modules/boltlib/lib/puppet/functions/run_command.rb [hunk 1]
```diff
@@ -26,7 +26,7 @@
   # @param options Additional options: '_catch_errors', '_run_as'.
   # @return A list of results, one entry per target.
   # @example Run a command on targets
-  #   run_command('hostname', $targets, '_catch_errors' => true)
+  #   run_command('hostname', $targets, 'Get hostname')
```
Comment (lucywyman) [inline bolt-modules/boltlib/lib/puppet/functions/run_command.rb]: Not sure I follow why this changed?

File: spec/bolt/executor_spec.rb [hunk 2]
```diff
@@ -310,6 +310,44 @@ def mock_node_results(_run_as = nil)
...
+      results = executor.wait_until_available(targets)
+      results.each do |result|
+        expect(result).to be_instance_of(Bolt::Result)
+      end
+    end
+
+    it 'errors after timeout' do
+      allow(ssh).to receive(:connected?).and_return(false)
+
+      results = executor.wait_until_available(targets, wait_time: 0, retry_interval: 0)
```
Comment (lucywyman) [inline spec/bolt/executor_spec.rb]: Might also be worth adding a test to ensure that retry_interval and wait_time are respected. wait_time would be easy to test, retry_interval might be trickier...you could maybe count how many times `Time.now` is called?

Comment (MikaelSmith) [reply spec/bolt/executor_spec.rb]: I can work on it. Getting the `wait` tests reliable has proved tricky.

File: lib/bolt/error.rb [hunk 3]
```diff
@@ -50,13 +50,14 @@ def initialize(msg)
...
 
-    def initialize(result_set, action, object)
+    def initialize(result_set, action, object = nil)
       details = {
         'action' => action,
         'object' => object,
         'result_set' => result_set
       }
-      message = "Plan aborted: #{action} '#{object}' failed on #{result_set.error_set.length} nodes"
+      object_msg = " '#{object}'" if object
```
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/bolt/error.rb]: It allows `object` to be optional.

File: bolt-modules/boltlib/lib/puppet/functions/run_command.rb (hunk 1 above)
Comment (MikaelSmith) [reply bolt-modules/boltlib/lib/puppet/functions/run_command.rb]: The example didn't map to the function, which takes a 3rd `description` argument before options.

File: spec/bolt/executor_spec.rb (hunk 2 above)
Comment (MikaelSmith) [reply spec/bolt/executor_spec.rb]: Added a new test and improved existing ones around `Time.now`.

**Pull request merged**